server.go

```diff
@@ -290,37 +266,28 @@ func (ca *crateAdapter) handleRead(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
-func writesToCrateRequest(req *remote.WriteRequest) *crateRequest {
-	request := &crateRequest{
-		BulkArgs: make([][]interface{}, 0, len(req.Timeseries)),
+func writesToCrateRequest(req *prompb.WriteRequest) *crateWriteRequest {
+	request := &crateWriteRequest{
+		rows: make([]*crateRow, 0, len(req.Timeseries)),
 	}
-	request.Stmt = fmt.Sprintf(`INSERT INTO metrics ("labels", "labels_hash", "value", "valueRaw", "timestamp") VALUES (?, ?, ?, ?, ?)`)
 
 	for _, ts := range req.Timeseries {
 		metric := make(model.Metric, len(ts.Labels))
 		for _, l := range ts.Labels {
 			metric[model.LabelName(l.Name)] = model.LabelValue(l.Value)
 		}
+		fp := metric.Fingerprint().String()
 
 		for _, s := range ts.Samples {
-			args := make([]interface{}, 0, 5)
-			args = append(args, metric)
-			args = append(args, metric.Fingerprint().String())
-			// Convert to string to handle NaN/Inf/-Inf.
-			switch {
-			case math.IsInf(s.Value, 1):
-				args = append(args, "Infinity")
-			case math.IsInf(s.Value, -1):
-				args = append(args, "-Infinity")
-			default:
-				args = append(args, fmt.Sprintf("%f", s.Value))
-			}
-			// Crate.io can't handle full NaN values as required by Prometheus 2.0,
-			// so store the raw bits as an int64.
-			args = append(args, int64(math.Float64bits(s.Value)))
-			args = append(args, s.TimestampMs)
-
-			request.BulkArgs = append(request.BulkArgs, args)
+			request.rows = append(request.rows, &crateRow{
+				labels:     metric,
+				labelsHash: fp,
+				timestamp:  time.Unix(0, s.Timestamp*1e6).UTC(),
+				value:      s.Value,
+				// Crate.io can't handle full NaN values as required by Prometheus 2.0,
```

Comment:
We would like to fix this. What exactly does CrateDB not support? What does "full NaN" mean?

Reply:
This is a comment that already existed in `master`, but I'm sure it's about the difference between "normal" NaN values and signalling NaN values, see https://github.com/prometheus/prometheus/blob/master/pkg/value/value.go and https://en.wikipedia.org/wiki/NaN#Signaling_NaN.

Prometheus uses one type of signalling NaN to explicitly mark stale samples. For this to work, the underlying database has to support storing the exact bit pattern that is sent to it. Not sure if CrateDB can do that via Postgres now?